File: server/app/services/classroom.py
from __future__ import annotations

import logging
from datetime import datetime, timedelta, timezone

from app.db.connection import get_db

logger = logging.getLogger("nexus.classroom")


class ClassroomService:
    """Server-side classroom management.

    Privacy rules (enforced):
    - Teacher sees mastery data, NEVER raw learning events
    - Students never see each other's data
    - No ranking, no comparison, no "class average" visible to students
    - Only first names — no PII beyond that
    """

    # --- Classroom lifecycle ---

    async def create_classroom(self, name: str, teacher_profile_id: str) -> dict:
        """Create a new classroom session."""
        db = await get_db()
        import uuid

        classroom_id = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

        await db.execute(
            "INSERT INTO classrooms (id, name, teacher_profile_id, status, created_at, last_activity) "
            "VALUES (?, ?, ?, 'active', ?, ?)",
            (classroom_id, name.strip(), teacher_profile_id, now, now),
        )
        await db.commit()

        return {
            "id": classroom_id,
            "name": name.strip(),
            "teacher_profile_id": teacher_profile_id,
            "status": "active",
            "created_at": now,
            "last_activity": now,
            "students": [],
            "groups": [],
        }

    async def get_classroom(self, classroom_id: str) -> dict | None:
        db = await get_db()
        cur = await db.execute("SELECT * FROM classrooms WHERE id = ?", (classroom_id,))
        row = await cur.fetchone()
        if row is None:
            return None
        return dict(row)

    # --- Student management ---

    async def add_student(
        self,
        classroom_id: str,
        student_profile_id: str,
    ) -> dict:
        """Add a student to a classroom (uses profile name from profiles table)."""
        db = await get_db()

        # Verify classroom exists
        cr = await db.execute("SELECT * FROM classrooms WHERE id = ?", (classroom_id,))
        classroom = await cr.fetchone()
        if classroom is None:
            raise ValueError("Classroom not found")

        # Check capacity
        cnt_cur = await db.execute(
            "SELECT COUNT(*) FROM classroom_students WHERE classroom_id = ?",
            (classroom_id,),
        )
        count = (await cnt_cur.fetchone())[0]
        if count >= 30:
            raise ValueError("Classroom is full (max 30 students)")

        # Check duplicate
        dup_cur = await db.execute(
            "SELECT 1 FROM classroom_students WHERE classroom_id = ? AND student_profile_id = ?",
            (classroom_id, student_profile_id),
        )
        if await dup_cur.fetchone():
            raise ValueError("Student already in classroom")

        # Get student name from profiles (first name only)
        prof_cur = await db.execute(
            "SELECT name FROM profiles WHERE id = ?", (student_profile_id,)
        )
        prof = await prof_cur.fetchone()
        if prof is None:
            raise ValueError("Student profile not found")

        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
        await db.execute(
            "INSERT INTO classroom_students (classroom_id, student_profile_id, joined_at) "
            "VALUES (?, ?, ?)",
            (classroom_id, student_profile_id, now),
        )
        await db.execute(
            "UPDATE classrooms SET last_activity = ? WHERE id = ?",
            (now, classroom_id),
        )
        await db.commit()

        return {
            "classroom_id": classroom_id,
            "student_profile_id": student_profile_id,
            "name": prof["name"],
            "joined_at": now,
        }

    async def remove_student(self, classroom_id: str, student_profile_id: str) -> None:
        db = await get_db()
        cur = await db.execute(
            "DELETE FROM classroom_students WHERE classroom_id = ? AND student_profile_id = ?",
            (classroom_id, student_profile_id),
        )
        if cur.rowcount == 0:
            raise ValueError("Student not in classroom")

        # Also remove from groups
        await db.execute(
            "DELETE FROM classroom_group_members WHERE student_profile_id = ? "
            "AND group_id IN (SELECT id FROM classroom_groups WHERE classroom_id = ?)",
            (student_profile_id, classroom_id),
        )
        await db.commit()

    async def get_students(self, classroom_id: str) -> list[dict]:
        """Get student roster with mastery tier — NO raw events, NO ranking."""
        db = await get_db()
        cur = await db.execute(
            "SELECT cs.student_profile_id, p.name, p.mastery_tier, cs.joined_at "
            "FROM classroom_students cs "
            "JOIN profiles p ON cs.student_profile_id = p.id "
            "WHERE cs.classroom_id = ? "
            "ORDER BY p.name",
            (classroom_id,),
        )
        rows = await cur.fetchall()
        return [
            {
                "profile_id": r["student_profile_id"],
                "name": r["name"],
                "mastery_tier": r["mastery_tier"],
                "joined_at": r["joined_at"],
            }
            for r in rows
        ]

    # --- Groups ---

    async def create_group(
        self, classroom_id: str, name: str, student_ids: list[str]
    ) -> dict:
        db = await get_db()
        import uuid

        group_id = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

        await db.execute(
            "INSERT INTO classroom_groups (id, classroom_id, name, created_at) VALUES (?, ?, ?, ?)",
            (group_id, classroom_id, name.strip(), now),
        )

        for sid in student_ids:
            await db.execute(
                "INSERT OR IGNORE INTO classroom_group_members (group_id, student_profile_id) "
                "VALUES (?, ?)",
                (group_id, sid),
            )

        await db.commit()

        return {
            "id": group_id,
            "classroom_id": classroom_id,
            "name": name.strip(),
            "student_ids": student_ids,
        }

    async def get_groups(self, classroom_id: str) -> list[dict]:
        db = await get_db()
        cur = await db.execute(
            "SELECT * FROM classroom_groups WHERE classroom_id = ? ORDER BY name",
            (classroom_id,),
        )
        groups = []
        for row in await cur.fetchall():
            m_cur = await db.execute(
                "SELECT student_profile_id FROM classroom_group_members WHERE group_id = ?",
                (row["id"],),
            )
            members = [r["student_profile_id"] for r in await m_cur.fetchall()]
            groups.append({
                "id": row["id"],
                "name": row["name"],
                "student_ids": members,
            })
        return groups

    # --- Assignments ---

    async def assign_quest(
        self, classroom_id: str, quest_id: str, group_id: str | None = None
    ) -> dict:
        db = await get_db()
        import uuid

        assignment_id = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

        await db.execute(
            "INSERT INTO classroom_assignments (id, classroom_id, assignment_type, quest_id, group_id, assigned_at) "
            "VALUES (?, ?, 'quest', ?, ?, ?)",
            (assignment_id, classroom_id, quest_id, group_id, now),
        )
        await db.commit()

        return {
            "id": assignment_id,
            "classroom_id": classroom_id,
            "quest_id": quest_id,
            "group_id": group_id,
            "assigned_at": now,
        }

    async def assign_focus(
        self, classroom_id: str, skills: list[str], group_id: str | None = None
    ) -> dict:
        db = await get_db()
        import json
        import uuid

        assignment_id = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

        await db.execute(
            "INSERT INTO classroom_assignments (id, classroom_id, assignment_type, skills_json, group_id, assigned_at) "
            "VALUES (?, ?, 'focus', ?, ?, ?)",
            (assignment_id, classroom_id, json.dumps(skills), group_id, now),
        )
        await db.commit()

        return {
            "id": assignment_id,
            "classroom_id": classroom_id,
            "skills": skills,
            "group_id": group_id,
            "assigned_at": now,
        }

    # --- Dashboard data (aggregated, privacy-safe) ---

    async def get_dashboard(self, classroom_id: str) -> dict:
        """Teacher dashboard — aggregated mastery, NEVER raw events."""
        db = await get_db()

        # Classroom info
        cr = await db.execute("SELECT * FROM classrooms WHERE id = ?", (classroom_id,))
        classroom = await cr.fetchone()
        if classroom is None:
            raise ValueError("Classroom not found")

        students = await self.get_students(classroom_id)

        # Aggregated mastery per subject
        subject_mastery: dict[str, dict] = {}
        for student in students:
            m_cur = await db.execute(
                "SELECT skill_id, level FROM mastery WHERE profile_id = ?",
                (student["profile_id"],),
            )
            for row in await m_cur.fetchall():
                skill_id = row["skill_id"]
                subject = skill_id.split(".")[0] if "." in skill_id else skill_id
                if subject not in subject_mastery:
                    subject_mastery[subject] = {
                        "subject": subject,
                        "total_level": 0.0,
                        "count": 0,
                        "at_mastery": 0,
                        "below_mastery": 0,
                    }
                sm = subject_mastery[subject]
                sm["total_level"] += row["level"]
                sm["count"] += 1
                if row["level"] >= 0.7:
                    sm["at_mastery"] += 1
                else:
                    sm["below_mastery"] += 1

        # Compute averages
        subject_breakdown = []
        for sm in subject_mastery.values():
            avg = sm["total_level"] / sm["count"] if sm["count"] > 0 else 0
            subject_breakdown.append({
                "subject": sm["subject"],
                "average_mastery": round(avg, 2),
                "students_at_mastery": sm["at_mastery"],
                "students_below_mastery": sm["below_mastery"],
            })

        # Common gaps (>40% struggling)
        total = len(students)
        threshold = max(1, int(total * 0.4))
        common_gaps = [
            s["subject"]
            for s in subject_breakdown
            if s["students_below_mastery"] >= threshold
        ]

        # Top strengths (>60% mastered)
        strength_threshold = max(1, int(total * 0.6))
        top_strengths = [
            s["subject"]
            for s in subject_breakdown
            if s["students_at_mastery"] >= strength_threshold
        ]

        return {
            "classroom_id": classroom_id,
            "name": classroom["name"],
            "status": classroom["status"],
            "total_students": total,
            "subject_breakdown": subject_breakdown,
            "common_gaps": common_gaps,
            "top_strengths": top_strengths,
            "students": students,
        }

    # --- Reports ---

    async def get_class_report(self, classroom_id: str) -> dict:
        """Mastery report for the class — privacy-safe, no ranking."""
        dashboard = await self.get_dashboard(classroom_id)

        recommendations = []
        if dashboard["common_gaps"]:
            recommendations.append(
                f"Consider focusing on {dashboard['common_gaps'][0]} — many students need support."
            )
        if dashboard["top_strengths"]:
            recommendations.append(
                f"The class excels at {dashboard['top_strengths'][0]} — consider advancing to harder challenges."
            )

        return {
            "classroom_id": classroom_id,
            "classroom_name": dashboard["name"],
            "generated_at": datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S"),
            "total_students": dashboard["total_students"],
            "subject_breakdown": dashboard["subject_breakdown"],
            "common_gaps": dashboard["common_gaps"],
            "top_strengths": dashboard["top_strengths"],
            "recommendations": recommendations,
        }

    # --- Settings ---

    async def update_settings(
        self,
        classroom_id: str,
        time_limit_minutes: int | None = None,
        status: str | None = None,
    ) -> dict:
        db = await get_db()
        now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

        updates = {"last_activity": now}
        if time_limit_minutes is not None:
            updates["time_limit_minutes"] = time_limit_minutes
        if status is not None:
            if status not in ("active", "paused", "ended"):
                raise ValueError(f"Invalid status: {status}")
            updates["status"] = status

        set_clause = ", ".join(f"{k} = ?" for k in updates)
        values = list(updates.values()) + [classroom_id]
        await db.execute(
            f"UPDATE classrooms SET {set_clause} WHERE id = ?", values  # noqa: S608
        )
        await db.commit()

        return await self.get_classroom(classroom_id) or {}

File: nexus-core/src/systems/classroom.ts
// Classroom mode — teacher dashboard + student management (up to 30 students)
// Privacy: teacher sees mastery data, NEVER raw learning events.
// No ranking, no comparison, no "class average" visible to students.

import type { System } from '../ecs/system.js';
import type { World } from '../ecs/world.js';
import type { MasteryTier } from '../types/components.js';
import type {
  Classroom,
  ClassroomSettings,
  ClassroomStudent,
  StudentGroup,
  QuestAssignment,
  FocusAssignment,
  ClassProgress,
  StudentProgress,
  SubjectProgress,
  ClassEngagement,
  EngagementLevel,
  ClassMasteryReport,
  ScreenTimeOverview,
  TeacherVisibleData,
  CreateClassroomInput,
  CreateGroupInput,
} from '../types/classroom.js';
import { DEFAULT_CLASSROOM_SETTINGS } from '../types/classroom.js';

// --- Privacy enforcement ---

function clampMastery(value: number): number {
  return Math.max(0, Math.min(1, Math.round(value * 100) / 100));
}

function extractSubject(skillId: string): string {
  const dot = skillId.indexOf('.');
  return dot > 0 ? skillId.slice(0, dot) : skillId;
}

// --- Engagement computation (aggregated, no individual exposure in real-time) ---

function computeEngagementLevel(
  mastery: number,
  recentAttempts: number,
  recentSuccesses: number,
): EngagementLevel {
  if (recentAttempts === 0) return 'thinking';
  const successRate = recentSuccesses / recentAttempts;
  if (successRate >= 0.6 && mastery >= 0.4) return 'in_flow';
  if (successRate >= 0.3) return 'thinking';
  return 'struggling';
}

// --- Mastery record for internal tracking ---

interface InternalMasteryRecord {
  skillId: string;
  level: number;
  attempts: number;
  successes: number;
}

// --- ClassroomSystem ---

export class ClassroomSystem implements System {
  readonly name = 'classroom';
  readonly priority = 47;

  private classrooms: Map<string, Classroom> = new Map();
  private questAssignments: Map<string, QuestAssignment[]> = new Map();
  private focusAssignments: Map<string, FocusAssignment[]> = new Map();
  private studentMastery: Map<string, InternalMasteryRecord[]> = new Map();
  private studentActivity: Map<string, { activeMinutes: number; questsCompleted: number }> =
    new Map();

  update(_world: World, _dt: number): void {
    // Check time limits and auto-pause if needed
    for (const classroom of this.classrooms.values()) {
      if (classroom.status !== 'active') continue;
      if (classroom.settings.timeLimitMinutes === null) continue;

      const elapsed =
        (Date.now() - new Date(classroom.lastActivity).getTime()) / 60000;
      if (elapsed >= classroom.settings.timeLimitMinutes) {
        classroom.status = 'paused';
      }
    }
  }

  // --- Classroom lifecycle ---

  createClassroom(input: CreateClassroomInput): Classroom {
    if (!input.name || input.name.trim().length === 0) {
      throw new Error('Classroom name is required');
    }
    if (!input.teacherProfileId) {
      throw new Error('Teacher profile ID is required');
    }

    const id = `classroom-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    const now = new Date().toISOString();
    const settings: ClassroomSettings = {
      ...DEFAULT_CLASSROOM_SETTINGS,
      ...(input.settings ?? {}),
    };

    const classroom: Classroom = {
      id,
      name: input.name.trim(),
      teacherProfileId: input.teacherProfileId,
      students: [],
      groups: [],
      settings,
      status: 'active',
      createdAt: now,
      lastActivity: now,
    };

    this.classrooms.set(id, classroom);
    this.questAssignments.set(id, []);
    this.focusAssignments.set(id, []);

    return classroom;
  }

  getClassroom(classroomId: string): Classroom | undefined {
    return this.classrooms.get(classroomId);
  }

  // --- Student management ---

  addStudent(
    classroomId: string,
    profileId: string,
    name: string,
    masteryTier: MasteryTier = 'foundation',
  ): void {
    const classroom = this.requireClassroom(classroomId);

    if (classroom.students.length >= classroom.settings.maxStudents) {
      throw new Error(
        `Classroom is full (max ${classroom.settings.maxStudents} students)`,
      );
    }

    if (classroom.students.some((s) => s.profileId === profileId)) {
      throw new Error('Student already in classroom');
    }

    classroom.students.push({
      profileId,
      name,
      masteryTier,
      joinedAt: new Date().toISOString(),
      groupId: null,
    });

    this.studentMastery.set(profileId, []);
    this.studentActivity.set(profileId, { activeMinutes: 0, questsCompleted: 0 });
    classroom.lastActivity = new Date().toISOString();
  }

  removeStudent(classroomId: string, studentProfileId: string): void {
    const classroom = this.requireClassroom(classroomId);
    const idx = classroom.students.findIndex((s) => s.profileId === studentProfileId);
    if (idx < 0) throw new Error('Student not in classroom');

    classroom.students.splice(idx, 1);

    // Remove from any groups
    for (const group of classroom.groups) {
      const gi = group.studentIds.indexOf(studentProfileId);
      if (gi >= 0) group.studentIds.splice(gi, 1);
    }

    classroom.lastActivity = new Date().toISOString();
  }

  getStudents(classroomId: string): ClassroomStudent[] {
    return this.requireClassroom(classroomId).students;
  }

  // --- Groups ---

  createGroup(classroomId: string, input: CreateGroupInput): StudentGroup {
    const classroom = this.requireClassroom(classroomId);

    if (!input.name || input.name.trim().length === 0) {
      throw new Error('Group name is required');
    }

    // Validate all students exist in classroom
    for (const sid of input.studentIds) {
      if (!classroom.students.some((s) => s.profileId === sid)) {
        throw new Error(`Student ${sid} not in classroom`);
      }
    }

    const group: StudentGroup = {
      id: `group-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      name: input.name.trim(),
      studentIds: [...input.studentIds],
    };

    classroom.groups.push(group);

    // Update students' group references
    for (const sid of input.studentIds) {
      const student = classroom.students.find((s) => s.profileId === sid);
      if (student) student.groupId = group.id;
    }

    classroom.lastActivity = new Date().toISOString();
    return group;
  }

  getGroups(classroomId: string): StudentGroup[] {
    return this.requireClassroom(classroomId).groups;
  }

  // --- Assignments ---

  assignQuest(classroomId: string, questId: string, groupId?: string): QuestAssignment {
    const classroom = this.requireClassroom(classroomId);

    if (groupId) {
      if (!classroom.groups.some((g) => g.id === groupId)) {
        throw new Error('Group not found');
      }
    }

    const assignment: QuestAssignment = {
      id: `qa-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      classroomId,
      questId,
      groupId: groupId ?? null,
      assignedAt: new Date().toISOString(),
    };

    this.questAssignments.get(classroomId)!.push(assignment);
    classroom.lastActivity = new Date().toISOString();
    return assignment;
  }

  assignFocus(classroomId: string, skills: string[], groupId?: string): FocusAssignment {
    const classroom = this.requireClassroom(classroomId);

    if (skills.length === 0) {
      throw new Error('At least one skill is required');
    }

    if (groupId) {
      if (!classroom.groups.some((g) => g.id === groupId)) {
        throw new Error('Group not found');
      }
    }

    const assignment: FocusAssignment = {
      id: `fa-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
      classroomId,
      skills: [...skills],
      groupId: groupId ?? null,
      assignedAt: new Date().toISOString(),
    };

    this.focusAssignments.get(classroomId)!.push(assignment);
    classroom.lastActivity = new Date().toISOString();
    return assignment;
  }

  getQuestAssignments(classroomId: string): QuestAssignment[] {
    return this.questAssignments.get(classroomId) ?? [];
  }

  getFocusAssignments(classroomId: string): FocusAssignment[] {
    return this.focusAssignments.get(classroomId) ?? [];
  }

  // --- Mastery data ingestion (from server or local DB) ---

  updateStudentMastery(profileId: string, records: InternalMasteryRecord[]): void {
    this.studentMastery.set(profileId, records);
  }

  updateStudentActivity(
    profileId: string,
    activeMinutes: number,
    questsCompleted: number,
  ): void {
    this.studentActivity.set(profileId, { activeMinutes, questsCompleted });
  }

  // --- Progress (aggregated — teacher sees mastery, NEVER raw events) ---

  getClassProgress(classroomId: string): ClassProgress {
    const classroom = this.requireClassroom(classroomId);
    const masteryBySubject = new Map<string, SubjectProgress>();
    const gapCounts = new Map<string, number>();
    const strengthCounts = new Map<string, number>();

    let activeCount = 0;

    for (const student of classroom.students) {
      const records = this.studentMastery.get(student.profileId) ?? [];
      const activity = this.studentActivity.get(student.profileId);
      if (activity && activity.activeMinutes > 0) activeCount++;

      for (const record of records) {
        const subject = extractSubject(record.skillId);
        let sp = masteryBySubject.get(subject);
        if (!sp) {
          sp = { subject, averageMastery: 0, studentsAtOrAbove: 0, studentsBelow: 0, totalStudents: 0 };
          masteryBySubject.set(subject, sp);
        }
        sp.totalStudents++;
        sp.averageMastery += record.level;

        if (record.level >= 0.7) {
          sp.studentsAtOrAbove++;
          strengthCounts.set(subject, (strengthCounts.get(subject) ?? 0) + 1);
        } else {
          sp.studentsBelow++;
          gapCounts.set(subject, (gapCounts.get(subject) ?? 0) + 1);
        }
      }
    }

    // Finalize averages
    for (const sp of masteryBySubject.values()) {
      if (sp.totalStudents > 0) {
        sp.averageMastery = clampMastery(sp.averageMastery / sp.totalStudents);
      }
    }

    // Common gaps: subjects where more than 40% of students are below mastery
    const totalStudents = classroom.students.length;
    const threshold = Math.max(1, Math.floor(totalStudents * 0.4));
    const commonGaps = Array.from(gapCounts.entries())
      .filter(([_, count]) => count >= threshold)
      .sort((a, b) => b[1] - a[1])
      .map(([subject]) => subject);

    // Top strengths: subjects where more than 60% are at mastery
    const strengthThreshold = Math.max(1, Math.floor(totalStudents * 0.6));
    const topStrengths = Array.from(strengthCounts.entries())
      .filter(([_, count]) => count >= strengthThreshold)
      .sort((a, b) => b[1] - a[1])
      .map(([subject]) => subject);

    // Recommendations
    const recommendations: string[] = [];
    if (commonGaps.length > 0) {
      recommendations.push(
        `Consider focusing on ${commonGaps[0]} this week — many students need support.`,
      );
    }
    if (topStrengths.length > 0) {
      recommendations.push(
        `The class excels at ${topStrengths[0]} — consider advancing to more complex challenges.`,
      );
    }
    if (activeCount < totalStudents * 0.8 && totalStudents > 0) {
      recommendations.push(
        `${totalStudents - activeCount} students have low engagement. Consider checking in with them.`,
      );
    }

    return {
      totalStudents,
      activeStudents: activeCount,
      masteryBySubject,
      commonGaps,
      topStrengths,
      recommendations,
    };
  }

  // --- Individual student progress (teacher drill-down) ---

  getStudentProgress(classroomId: string, studentId: string): StudentProgress | null {
    const classroom = this.requireClassroom(classroomId);
    const student = classroom.students.find((s) => s.profileId === studentId);
    if (!student) return null;

    const records = this.studentMastery.get(studentId) ?? [];
    const activity = this.studentActivity.get(studentId) ?? {
      activeMinutes: 0,
      questsCompleted: 0,
    };

    const subjectMastery = new Map<string, number>();
    const subjectCounts = new Map<string, number>();
    let totalLevel = 0;

    for (const r of records) {
      const subject = extractSubject(r.skillId);
      subjectMastery.set(subject, (subjectMastery.get(subject) ?? 0) + r.level);
      subjectCounts.set(subject, (subjectCounts.get(subject) ?? 0) + 1);
      totalLevel += r.level;
    }

    // Average per subject
    for (const [subj, total] of subjectMastery) {
      const count = subjectCounts.get(subj) ?? 1;
      subjectMastery.set(subj, clampMastery(total / count));
    }

    return {
      profileId: studentId,
      name: student.name,
      masteryTier: student.masteryTier,
      overallMastery: records.length > 0 ? clampMastery(totalLevel / records.length) : 0,
      subjectMastery,
      questsCompleted: activity.questsCompleted,
      activeMinutes: activity.activeMinutes,
    };
  }

  // --- Teacher-visible data (privacy-safe) ---

  getTeacherVisibleData(classroomId: string): TeacherVisibleData[] {
    const classroom = this.requireClassroom(classroomId);
    const result: TeacherVisibleData[] = [];

    for (const student of classroom.students) {
      const records = this.studentMastery.get(student.profileId) ?? [];
      const activity = this.studentActivity.get(student.profileId) ?? {
        activeMinutes: 0,
        questsCompleted: 0,
      };

      const subjectMastery = new Map<string, number>();
      const subjectCounts = new Map<string, number>();
      let totalLevel = 0;
      let totalAttempts = 0;
      let totalSuccesses = 0;

      for (const r of records) {
        const subject = extractSubject(r.skillId);
        subjectMastery.set(subject, (subjectMastery.get(subject) ?? 0) + r.level);
        subjectCounts.set(subject, (subjectCounts.get(subject) ?? 0) + 1);
        totalLevel += r.level;
        totalAttempts += r.attempts;
        totalSuccesses += r.successes;
      }

      for (const [subj, total] of subjectMastery) {
        const count = subjectCounts.get(subj) ?? 1;
        subjectMastery.set(subj, clampMastery(total / count));
      }

      const overallMastery = records.length > 0 ? clampMastery(totalLevel / records.length) : 0;
      const engagement = computeEngagementLevel(overallMastery, totalAttempts, totalSuccesses);

      result.push({
        profileId: student.profileId,
        name: student.name,
        masteryTier: student.masteryTier,
        overallMastery,
        subjectMastery,
        questsCompleted: activity.questsCompleted,
        engagementLevel: engagement,
      });
    }

    return result;
  }

  // --- Aggregated engagement (no individual struggles exposed) ---

  getClassEngagement(classroomId: string): ClassEngagement {
    const data = this.getTeacherVisibleData(classroomId);
    let inFlow = 0;
    let thinking = 0;
    let struggling = 0;

    for (const d of data) {
      switch (d.engagementLevel) {
        case 'in_flow':
          inFlow++;
          break;
        case 'thinking':
          thinking++;
          break;
        case 'struggling':
          struggling++;
          break;
      }
    }

    return {
      totalStudents: data.length,
      inFlow,
      thinking,
      struggling,
    };
  }

  // --- Reports ---

  generateMasteryReport(classroomId: string): ClassMasteryReport {
    const classroom = this.requireClassroom(classroomId);
    const progress = this.getClassProgress(classroomId);

    const subjectBreakdown = Array.from(progress.masteryBySubject.values());
    const overallMastery =
      subjectBreakdown.length > 0
        ? clampMastery(
            subjectBreakdown.reduce((acc, s) => acc + s.averageMastery, 0) /
              subjectBreakdown.length,
          )
        : 0;

    // Screen time overview
    let totalMinutes = 0;
    let totalSessions = 0;
    for (const student of classroom.students) {
      const activity = this.studentActivity.get(student.profileId);
      if (activity) {
        totalMinutes += activity.activeMinutes;
        totalSessions++;
      }
    }

    const screenTimeOverview: ScreenTimeOverview = {
      averageMinutes:
        totalSessions > 0 ? Math.round(totalMinutes / totalSessions) : 0,
      totalSessions,
    };

    return {
      classroomId,
      classroomName: classroom.name,
      generatedAt: new Date().toISOString(),
      totalStudents: classroom.students.length,
      overallMastery,
      subjectBreakdown,
      commonGaps: progress.commonGaps,
      topStrengths: progress.topStrengths,
      recommendations: progress.recommendations,
      screenTimeOverview,
    };
  }

  // --- Classroom controls ---

  pauseClassroom(classroomId: string): void {
    const classroom = this.requireClassroom(classroomId);
    classroom.status = 'paused';
    classroom.lastActivity = new Date().toISOString();
  }

  resumeClassroom(classroomId: string): void {
    const classroom = this.requireClassroom(classroomId);
    if (classroom.status === 'ended') {
      throw new Error('Cannot resume an ended classroom');
    }
    classroom.status = 'active';
    classroom.lastActivity = new Date().toISOString();
  }

  endClassroom(classroomId: string): void {
    const classroom = this.requireClassroom(classroomId);
    classroom.status = 'ended';
    classroom.lastActivity = new Date().toISOString();
  }

  updateSettings(classroomId: string, updates: Partial<ClassroomSettings>): ClassroomSettings {
    const classroom = this.requireClassroom(classroomId);
    Object.assign(classroom.settings, updates);
    classroom.lastActivity = new Date().toISOString();
    return { ...classroom.settings };
  }

  // --- Internal helpers ---

  private requireClassroom(classroomId: string): Classroom {
    const classroom = this.classrooms.get(classroomId);
    if (!classroom) throw new Error('Classroom not found');
    return classroom;
  }

  // --- Cleanup ---

  reset(): void {
    this.classrooms.clear();
    this.questAssignments.clear();
    this.focusAssignments.clear();
    this.studentMastery.clear();
    this.studentActivity.clear();
  }
}
